### PR TITLE
[ty] make `StaticClassLiteral::explicit_bases` converge better in cycles

### PR DESCRIPTION
**Pull request by oconnor663**
Fixes https://github.com/astral-sh/ty/issues/2637. This PR adds a minimized test case derived from that issue. It creates [a complicated cycle with multiple cycle heads](https://gist.github.com/oconnor663/c996ed2cc97d172dd4b9a8d8207dc7ac), which also involves [a tricky Salsa behavior that comes up when a query oscillates between being a cycle head and not being one](https://gist.github.com/oconnor663/c2a7662e3d88048b691754da957121d1).

The fix here is to improve the cycle behavior of `StaticClassLiteral::explicit_bases`. Previously its `cycle_initial` returned an empty list, but an empty list of bases causes assignability checks to fail, where a temporarily-`Divergent` base class would succeed. Returning a list of `Divergent` bases behaves better. The new `cycle_fn` wasn't necessary to fix this case, but it seemed like a good idea to also make sure these types enjoy the same "monotonic widening" that we apply to avoid oscillations in other places.

### CODE REVIEW

File: crates/ty_python_semantic/resources/mdtest/cycle.md
Comment (AlexWaygood) [inline crates/ty_python_semantic/resources/mdtest/cycle.md:1]: It would be great to add some `reveal_type` and `reveal_mro` calls to this test, just to check that none of the `Divergent` types still remain after the cycle has been resolved, e.g.

````diff
diff --git a/crates/ty_python_semantic/resources/mdtest/cycle.md b/crates/ty_python_semantic/resources/mdtest/cycle.md
index c8c72640f4..031a8c41b0 100644
--- a/crates/ty_python_semantic/resources/mdtest/cycle.md
+++ b/crates/ty_python_semantic/resources/mdtest/cycle.md
@@ -170,12 +170,14 @@ which also involved
 ```py
 from derived import Derived
 
-Derived.decorate
+# revealed: bound method <class 'Derived'>.decorate[T](item_class: type[T]) -> type[T]
+reveal_type(Derived.decorate)
 ```
 
 `derived.py`:
 
 ```py
+from ty_extensions import reveal_mro
 import bases
 
 class Derived(bases.GenericBase["Foo", "Bar"]): ...
@@ -183,14 +185,25 @@ class Derived(bases.GenericBase["Foo", "Bar"]): ...
 @Derived.decorate
 class Foo(bases.Foo): ...
 
+# revealed: <class 'Foo'>
+reveal_type(Foo)
+# revealed: (<class 'derived.Foo'>, <class 'bases.Foo'>, <class 'object'>)
+reveal_mro(Foo)
+
 @Derived.decorate
 class Bar(bases.Bar): ...
+
+# revealed: <class 'Bar'>
+reveal_type(Bar)
+# revealed: (<class 'derived.Bar'>, <class 'bases.Bar'>, <class 'object'>)
+reveal_mro(Bar)
 ```
 
 `bases.py`:
 
 ```py
 from typing import Generic, TypeVar, Type
+from ty_extensions import reveal_mro
 
 T = TypeVar("T")
 B1 = TypeVar("B1", bound="Foo")
@@ -201,6 +214,13 @@ class GenericBase(Generic[B1, B2]):
     def decorate(cls, item_class: Type[T]) -> Type[T]:
         return item_class
 
+# revealed: <class 'GenericBase'>
+reveal_type(GenericBase)
+# revealed: (<class 'GenericBase[Unknown, Unknown]'>, typing.Generic, <class 'object'>)
+reveal_mro(GenericBase)
+# revealed: (<class 'GenericBase[Foo, Bar]'>, typing.Generic, <class 'object'>)
+reveal_mro(GenericBase["Foo", "Bar"])
+
 class Foo: ...
 class Bar: ...
````

All the assertions appear to pass just fine!
